Fix HostPortOrFile to support IPv6 addresses with zone

bcrochet opened the pull request:
1. The HostPortOrFile tests don't have any IPv6 tests. This adds some.
2. The HostPortOrFile breaks if any of the addresses have IPv6 zone
defined. ParseIP does not handle %zone anymore.

<!--
Thank you for contributing to CoreDNS!
Please provide the following information to help us make the most of your pull request:
-->

### 1. Why is this pull request needed and what does it do?

This was discovered using the forward plugin, but perhaps would affect others. The HostPortOrFile function does not properly support an IPv6 address of the form address%zone. 

### 2. Which issues (if any) are related?
None

### 3. Which documentation changes (if any) need to be made?
None
### 4. Does this introduce a backward incompatible change or deprecation?
No

miekg commented on plugin/pkg/parse/host.go:
```diff
@@ -24,11 +24,17 @@ func HostPortOrFile(s ...string) ([]string, error) {
-				// Not an IP address.
-				ss, err := tryFile(host)
+				// Not an IPv4 address, for sure. Could be IPv6 with %zone?
+				addrs, err := net.LookupHost(host)
```
you can't use `net.LookupHost` here because you may not have a resolver at your disposal

bcrochet replied:
I was afraid of that. I wasn't totally sure that this was the correct approach, but it seemed like a good first step.

bcrochet commented:
Filed bug #3528

miekg commented:
I won't have to time to work on this myself. But iirc the go std library
should be able to parse and see the zone part. Tricky bit might be
separating from a weirdly named file

On Wed, 11 Dec 2019, 13:50 Brad P. Crochet, <notifications@github.com>
wrote:

> *@bcrochet* commented on this pull request.
>
> doesn't net.ParseIP do this correct but itself?
>
> In general, I think this is a good change.
>
> That's why I'm trying to figure this out. ParseIP itself is not
> sufficient. If you pass ParseIP an IP with a %zone, it returns nil. So,
> while that isn't the worst thing, it just means that there needs to be more
> checks. Feel free to try running the tests with the additional cases that
> I've added, without the net.LookupHost. The ones with the %zone will fail.
>
> —
> You are receiving this because your review was requested.
> Reply to this email directly, view it on GitHub
> <https://github.com/coredns/coredns/pull/3527?email_source=notifications&email_token=AACWIWZ3AX2R4I2EYDOBDFTQYDVZRA5CNFSM4JY6SBTKYY3PNVWWK3TUL52HS4DFWFIHK3DMKJSXC5LFON2FEZLWNFSXPKTDN5WW2ZLOORPWSZGOCOZ5DNA#pullrequestreview-330551732>,
> or unsubscribe
> <https://github.com/notifications/unsubscribe-auth/AACWIW4WYN3K6X34J6WAPY3QYDVZRANCNFSM4JY6SBTA>
> .
>

bcrochet commented:
net.ParseIP definitely does not handle an address with %zone in it. I added the test cases with IPv6 addresses first, and ran them before I added any code. The ones with %zone failed.

miekg commented:
I think you need to net.ParseIP trimming off the `%scope` suffix. If that's an IP `%scope` is a scope. Note sure if `scope` can be syntax checked.

(also note I said zone, but it is scope before)

bcrochet commented:
If I'm understanding correctly, you're saying to trim off %scope before passing to ParseIP?

miekg commented:
[ Quoting <notifications@github.com> in "Re: [coredns/coredns] Fix HostPortO..." ]
>If I'm understanding correctly, you're saying to trim off %scope before passing to ParseIP?

yep. Makes sense, no?

ironcladlou commented:
@miekg looks like the zone stripping feedback was addressed, PTAL? Thanks!
